webkit2/webkit2.uri-scheme-request.lisp

```diff
@@ -0,0 +1,105 @@
+    ())
+
+(defcfun "webkit_uri_scheme_request_get_scheme" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-scheme)
+
+(defcfun "webkit_uri_scheme_request_get_uri" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-uri)
+
+(defcfun "webkit_uri_scheme_request_get_path" :string
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-path)
+
+(defcfun "webkit_uri_scheme_request_get_web-view" (g-object webkit-web-view)
+  (request (g-object webkit-uri-scheme-request)))
+(export 'webkit-uri-scheme-request-get-web-view)
+
+(defcfun "webkit_uri_scheme_request_finish" :void
+  (request (g-object webkit-uri-scheme-request))
+  (stream (g-object g-memory-input-stream))
+  (stream-length :long)
+  (content-type :string))
+(export 'webkit-uri-scheme-request-finish)
+
+(defcfun ("webkit_uri_scheme_request_finish_error" %webkit-uri-scheme-request-finish-error) :void
+  (request (g-object webkit-uri-scheme-request))
+  (g-error :pointer))
+
+(defun webkit-uri-scheme-request-finish-error (request error-string)
+  (%webkit-uri-scheme-request-finish-error
+   request (glib::%g-error-new-literal
+            +webkit-plugin-error+
+            4
+            error-string)))
+(export 'webkit-uri-scheme-request-finish-error)
+
+(defcfun "g_memory_input_stream_new_from_data" (g-object g-memory-input-stream)
+  (data :pointer)
+  (length :long)
+  (destroy :pointer))
+
+(cffi:defcallback g-notify-destroy-null :void ((data :pointer))
+  (declare (ignore data)))
+(cffi:defcallback g-notify-destroy-free :void ((data :pointer))
+  (cffi:foreign-funcall "free" :pointer data))
+
+(cffi:defcallback uri-scheme-processed :void ((request (g-object webkit-uri-scheme-request))
+                                              (user-data :pointer))
+  (let ((callback (find (cffi:pointer-address user-data) callbacks :key (function callback-id))))
+    (when (callback-function callback)
+      ;; Callback function returns data-string as a first value
+      ;; and data type (e.g., "text/html") as an optional second
+      ;; value.
+      (destructuring-bind (data &optional (data-type "text-html"))
+          (multiple-value-list (funcall (callback-function callback) request))
+        (handler-case
+            (multiple-value-bind (ffi-string ffi-string-length)
+                (cffi:foreign-string-alloc data)
+              (let* ((stream (g-memory-input-stream-new-from-data
+                              ffi-string ffi-string-length (callback g-notify-destroy-null))))
+                (webkit-uri-scheme-request-finish request stream ffi-string-length data-type)
+                (gobject:g-object-unref (pointer stream))
+                (cffi:foreign-string-free ffi-string)))
+          (error (c)
+            (webkit-uri-scheme-request-finish-error
+             request (format nil "The custom url request for URI ~S failed with ~A: ~A"
```

Comment:
Nit: We usually use lower case format characters, i.e. `~s` and `~a`, no?  Not sure what's preferred.

Reply:
I believe they have different meaning based on case.

Reply:
I think not, the CLHS page is uncased: http://www.lispworks.com/documentation/HyperSpec/Body/22_cda.htm.

Reply:
Downcased!